### PR TITLE
DEV-1386 Link to datatype/sample once per file

### PR DESCRIPTION
**Pull request by nedleitch**
Not precisely sure yet why sometimes multiple copies of the
same operation are requested but it is not incorrect to store
and apply the operations in a set rather than a list. 

This will resolve the problem we are currently seeing 
anyway.

### CODE REVIEW

File: cluster/src/main/java/com/hartwig/pipeline/transfer/SbpFileApiUpdate.java [hunk 1]
```diff
@@ -30,14 +30,15 @@
     private final SbpRun sbpRun;
     private final Bucket sourceBucket;
     private final SbpRestApi sbpApi;
-    private final List<ApiFileOperation> fileOperations = new ArrayList<>();
+    private final Set<ApiFileOperation> fileOperations;
```
Comment (pauldwolfe) [inline cluster/src/main/java/com/hartwig/pipeline/transfer/SbpFileApiUpdate.java:33]: As neither implementation of ApiFileOperation implement equals/hashcode this is only ensuring that the same reference isn't in the collection twice, which feels like masking another bug. I think we should keep going until we find where that is happening.

Still fine to keep idea but maybe then better we check for duplicates and log something before removing them?

File: cluster/src/main/java/com/hartwig/pipeline/transfer/SbpFileApiUpdate.java [hunk 2]
```diff
@@ -59,14 +62,26 @@ public void accept(final Blob blob) {
...
-                        LOGGER.info("Applying: {}", fileOperation);
+                        LOGGER.info("Applying [{}] for [{}]", fileOperation, blob.getName());
                         fileOperation.apply(sbpApi, fileResponse);
                     }
                 }
             }
         }
     }
 
+    private void storeOperations(List<StageOutput> stageOutputs) {
```
Comment (pauldwolfe) [inline cluster/src/main/java/com/hartwig/pipeline/transfer/SbpFileApiUpdate.java:73]: I would suggest reverting all this to get back to a place where you can duplicate the bug. I don't see a possible concurrency issue as the new code only processes the list which should already contain the duplicate. If you can't duplicate it after that, it suggests we still don't quite understand the problem.

File: cluster/src/main/java/com/hartwig/pipeline/PipelineState.java [hunk 3]
```diff
@@ -13,7 +13,15 @@
...
 
     private static final Logger LOGGER = LoggerFactory.getLogger(PipelineState.class);
-    private final List<StageOutput> stageOutputs = Lists.newArrayList();
+    private final List<StageOutput> stageOutputs;
+
+    public PipelineState() {
+        this.stageOutputs = new ArrayList<>();
+    }
+
+    public PipelineState(final List<StageOutput> stageOutputs) {
```
Comment (nedleitch) [inline cluster/src/main/java/com/hartwig/pipeline/PipelineState.java:22]: [Package?] private?

File: cluster/src/main/java/com/hartwig/pipeline/transfer/SbpFileApiUpdate.java (hunk 2 above)
Comment (pauldwolfe) [reply cluster/src/main/java/com/hartwig/pipeline/transfer/SbpFileApiUpdate.java:73]: I don't think we should keep this code, its a little confusing when you don't have the full context of why its here. If we do keep, the second info statement should probably be a WARN.

**Pull request merged**